cranelift/codegen/src/isa/riscv64/lower.isle

```diff
@@ -927,17 +918,35 @@
 
 
 ;;;;;  Rules for `select_spectre_guard`;;;;;;;;;
-(rule
-  (lower (has_type r_ty (select_spectre_guard (icmp cc ca @ (value_type cty) cb) a b)))
-  (let
-    ((dst VecWritableReg (alloc_vec_writable r_ty))
-      (r Reg (lower_icmp cc ca cb cty))
-      (_ Unit (emit (MInst.SelectIf $true (vec_writable_clone dst) r a b))))
-    (vec_writable_to_regs dst)))
 
-(rule -1
-  (lower (has_type ty (select_spectre_guard c @ (value_type cty) x y)))
-  (gen_select ty (truthy_to_reg cty (normalize_cmp_value cty c (ExtendOp.Zero))) x y))
+;; SelectSpectreGuard is equivalent to Select, but we should not use a branch based
+;; lowering for it. Instead we use a conditional move based lowering.
+;;
+;; We don't have cmov's in RISC-V either, but we can emulate those using bitwise
+;; operations, which is what we do below.
+(rule (lower (has_type ty (select_spectre_guard cmp @ (value_type cty) x @ (value_type arg_ty) y)))
+  (let ((normalized Reg (truthy_to_reg cty (normalize_cmp_value cty cmp (ExtendOp.Zero))))
+        ;; Build a value that is 0 or 1 depending on if `normalized` is zero.
+        ;;
+        ;; Translates roughly into the following rust expression:
+        ;; non_zero = ((normalized | ((!normalized) + 1)) >> 64) & 1
+        (a Reg (rv_addi (rv_not normalized) (imm12_const 1)))
+        (b Reg (rv_or normalized a))
+        (c Reg (rv_srli b (imm12_const 63)))
+        (non_zero Reg (rv_andi c (imm12_const 1)))
+        ;; Based on the 0 or 1 value we now build a mask that is either 0 or -1
+        (mask Reg (rv_addi non_zero (imm12_const -1)))
```

Comment:
Is the reason we can't use `lower_bmask` here its ultimate use of `gen_select_reg`? Would it be bad to rework the lowering of `SelectReg` to take this approach instead?

Reply:
I hadn't even considered that this was a `bmask` operation! And yeah, I think this is a better lowering than what we currently have. However I check what [LLVM produces for `bmask`](https://godbolt.org/z/dcqq34d3d) and It's slightly better, so I'm going to rework `bmask` to use that and use that here.

Reply:
I'm much happier with the output of this now, Thanks!

---

I'm now re-reading your comment, and realizing there was a second part about `SelectReg` that I missed. 

I have somewhat mixed feelings about making `Inst::SelectReg` do something like this, It would be quite a bit longer (~4 instructions I think?). But it would be really nice since we could move it entirely into ISLE. However I don't think that justifies it, it's a really common operation in the backend. 

I'm willing to be persuaded otherwise though!